src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,146 @@
+use Doctrine\DBAL\Types\Type;
+
+/**
+ * @author Vincent Touzet <vincent.touzet@gmail.com>
+ */
+class Connection
+{
+    private $tableName;
+    private $redeliverTimeout;
+    private $debug;
+    private $driverConnection;
+
+    public function __construct(string $tableName, DriverConnection $driverConnection, bool $debug = false, int $redeliverTimeout = 3600)
+    {
+        $this->tableName = $tableName;
+        $this->redeliverTimeout = $redeliverTimeout;
+        $this->driverConnection = $driverConnection;
+        $this->debug = $debug;
+    }
+
+    public function publish(string $body, array $headers): void
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+        $this->driverConnection->insert($this->tableName, [
+            'body' => $body,
+            'headers' => \json_encode($headers),
+            'created_at' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+            'available_at' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+        ]);
+    }
+
+    public function get(): ?array
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $this->driverConnection->beginTransaction();
+        try {
+            $query = $this->driverConnection->createQueryBuilder()
+                ->select('m.*')
+                ->from($this->tableName, 'm')
+                ->where('m.delivered_at is null')
+                ->orderBy('available_at', 'ASC')
+                ->setMaxResults(1);
+
+            // use SELECT ... FOR UPDATE to lock table
+            $doctrineEnvelope = $this->driverConnection->executeQuery(
+                $query->getSQL().' '.$this->driverConnection->getDatabasePlatform()->getWriteLockSQL()
+            )->fetch();
+
+            if (false === $doctrineEnvelope) {
+                $this->driverConnection->commit();
+
+                return null;
+            }
+
+            $doctrineEnvelope['headers'] = \json_decode($doctrineEnvelope['headers'], true);
+
+            $this->driverConnection->createQueryBuilder()
+                ->update($this->tableName)
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id')
+                ->setParameter('delivered_at', (new \DateTime('now'))->format('Y-m-d H:i:s'))
+                ->setParameter('id', $doctrineEnvelope['id'])
+                ->execute();
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Exception $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->tableName, ['id' => $id]) > 0;
+    }
+
+    public function nack($id): bool
+    {
+        return $this->driverConnection->update($this->tableName, ['delivered_at' => null], ['id' => $id]) > 0;
+    }
+
+    public function requeueMessages()
```

Comment:
I don't really understand this method. Why doesn't it take a specific message ID that fails? Why isn't it setting the `available_at` field?

Reply:
This just allow to requeue messages that are still considered as being handled for a long time (default to 3600). This method is called just before retrieving the next message to handle.

The use case is to requeue messages for handler that failed (server crash, Doctrine connection error, ...). This will requeue the message (i.e: setting delivered_at to null so it can be retrieved by a handler)